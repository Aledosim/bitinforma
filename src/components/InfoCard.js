import React, { useContext, useEffect, useState } from 'react'

import { CurrencyContext } from '../contexts/CurrencyContext'

import coinLogo from '../images/coins/btc.svg'

import styles from '../styles/components/InfoCard.module.css'

export default function InfoCard() {

    const {
        vol,
        closing,
        volBRL,
        currencyName
    } = useContext(CurrencyContext)

    const [volOut, setVolOut] = useState()
    const [closingOut, setClosingOut] = useState()
    const [volBRLOut, setVolBRLOut] = useState()

    useEffect(() => volHandler(vol), [vol])
    useEffect(() => closingHandler(closing), [closing])
    useEffect(() => volBRLHandler(volBRL), [volBRL])

    function volHandler(rawVol) {
        rawVol = Math.trunc(rawVol * 100) / 100

        var volInt
        var volDecimal
        [ volInt, volDecimal ] = String(rawVol).split('.')

        if (typeof(volDecimal) == 'undefined') {
            volDecimal = '0'
        }
        setVolOut(volInt + ',' + volDecimal.padEnd(2, '0'))
    }

    function closingHandler(rawClosing) {
        rawClosing = Math.round(rawClosing) / 1000

        var closingInt
        var closingDecimal
        [ closingInt, closingDecimal ] = String(rawClosing).split('.')

        if (typeof(closingDecimal) == 'undefined') {
            closingDecimal = '0'
        }

        setClosingOut(closingInt + '.' + closingDecimal.padEnd(3, '0'))
    }

    function volBRLHandler(rawVol) {
        const vol = rawVol.toLocaleString('pt-br', {
            style: 'currency',
            currency: 'BRL',
            minimumFractionDigits: 2
        })
        setVolBRLOut(vol)
    }

    return(
        <div id='infocard' className={styles.infoCardContainer}>
            <header className={styles.header}>
                <div className={styles.logoContainer}>
                    <span>
                        <img src={coinLogo} alt='coin logo'></img>
                    </span>
                    <span>
                        {currencyName}
                    </span>
                </div>
                <div id='price' className={styles.price}>
                    R$ {closingOut}
                </div>
            </header>
            <div>
                <p id='volume'>
                    {volOut} bitcoins negociados nas Ãºltimas 24hs
                </p>
                <p id='total'>
                    Um total de {volBRLOut}
                </p>
                <div>
                    <a href='https://www.mercadobitcoin.com.br/negociacoes'>Veja mais -></a>
                </div>
            </div>
        </div>
    )
}

