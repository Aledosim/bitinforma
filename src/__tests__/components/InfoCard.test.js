import { enableFetchMocks  } from 'jest-fetch-mock'

enableFetchMocks()
import React from 'react'
import { render, screen, waitFor } from '@testing-library/react'

import { tickerResponse, summaryResponse } from '../fixtures.json'
import CurrencyProvider from '../../contexts/CurrencyContext'

import InfoCard, { volHandler, closingHandler, volBRLHandler } from '../../components/InfoCard'

describe('volHandler tests', () => {
  it('should return a function', () => {
    const setVolOut = jest.fn()

    const func = volHandler(setVolOut)

    expect(typeof func).toEqual('function')
  });

  it('should return the formated sting', () => {
    const func = volHandler()
    const result = func('159.91613864')

    expect(result).toEqual("159,91")
  });

});

describe('closingHandler tests', () => {
  it('should return a function', () => {
    const func = closingHandler()

    expect(typeof func).toEqual('function')
  });

  it('should return the sting formated value', () => {
    const func = closingHandler()
    const result = func('326900.00666999')

    expect(result).toEqual("326.900")
  });

  it('should return the correct value if the input ends with 0', () => {
    const func = closingHandler()
    const result = func('326000.00666999')

    expect(result).toEqual("326.000")
  });

});

describe('volBRLHandler tests', () => {
  it('should return a function', () => {
    const func = volBRLHandler()

    expect(typeof func).toEqual('function')
  });

  it('should return the string formated value', () => {
    const func = volBRLHandler()
    const result = func(41198719.10154287)

    expect(result).toEqual("R$Â 41.198.719,10")
  });

});

describe('<InfoCard /> tests', () => {
  beforeEach(() => {
    fetch.resetMocks()

    fetch
      .mockResponse(req => {
        if (/.*\/ticker\/.*/.test(req.url)) {
          return Promise.resolve({ body: JSON.stringify(tickerResponse) })

        } else if (/.*\/day-summary\/.*/.test(req.url)) {
          return Promise.resolve({ body: JSON.stringify(summaryResponse) })

        }
      })
  })

  it('renders without crashing', async () => {

    const tree = render(
      <CurrencyProvider>
        <InfoCard />
      </CurrencyProvider>
    )

    // Check if the fetch data is rendered
    await waitFor(() => {
      expect(fetch).toHaveBeenCalledTimes(2)
    })

    expect(tree).toMatchSnapshot()

  })
})

