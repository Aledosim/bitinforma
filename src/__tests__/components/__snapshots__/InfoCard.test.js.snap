// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<InfoCard /> tests renders without crashing 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="infoCardContainer"
        data-cy="infocard"
      >
        <header
          class="header"
        >
          <div
            class="logoContainer"
            data-cy="coinlogo"
          >
            <span>
              <img
                alt="coin logo"
                src="btc.svg"
              />
            </span>
            <span>
              Bitcoin
            </span>
          </div>
          <div
            class="price"
            data-cy="price"
            data-testid="price"
          >
            R$ 
            326.900
          </div>
        </header>
        <div>
          <p
            id="volume"
          >
            259,88
             bitcoins negociados nas últimas 24hs
          </p>
          <p
            id="total"
          >
            Um total de 
            R$ 41.198.719,10
          </p>
          <div>
            <a
              href="https://www.mercadobitcoin.com.br/negociacoes"
            >
              Veja mais -&gt;
            </a>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="infoCardContainer"
      data-cy="infocard"
    >
      <header
        class="header"
      >
        <div
          class="logoContainer"
          data-cy="coinlogo"
        >
          <span>
            <img
              alt="coin logo"
              src="btc.svg"
            />
          </span>
          <span>
            Bitcoin
          </span>
        </div>
        <div
          class="price"
          data-cy="price"
          data-testid="price"
        >
          R$ 
          326.900
        </div>
      </header>
      <div>
        <p
          id="volume"
        >
          259,88
           bitcoins negociados nas últimas 24hs
        </p>
        <p
          id="total"
        >
          Um total de 
          R$ 41.198.719,10
        </p>
        <div>
          <a
            href="https://www.mercadobitcoin.com.br/negociacoes"
          >
            Veja mais -&gt;
          </a>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
